### PR TITLE
[WIP] Fix for #4139 - Secrets not being masked in pack config

### PR DESCRIPTION
**Pull request by nmaludy**
Fix for #4139 

This adds recursion into the secret masking process so that configs with nested `dicts` and `lists` will also have their secrets masked out for API calls.

**TODO**
- [] Unit tests
- [] Test various combinations of dicts/lists/other
- [] Find other edge cases

### CODE REVIEW

**Comment by nmaludy** [comment]
@Kami Here's an initial workup of a fix. Initial testing looks good (used the email pack schema as a reference https://github.com/StackStorm-Exchange/stackstorm-email/blob/master/config.schema.yaml).

I still need to develop test cases, but would appreciate comments early if you have time.

**Comment by Kami** [comment]
Thanks, I will have a look tomorrow.

And yeah, I agree - we need a good and thorough test coverage for that.

Another thing we need to keep in mind is to try to make this function as performance efficient as possible (bail early if there are no secret params, only iterate over items which could contain secret params based on the schema, etc.).

If we don't do that and just blindly recursively iterate over nested objects, it has a potential to significantly slow API responses for all the endpoints with deeply nested objects (executions, etc.).

**Comment by nmaludy** [comment]
An idea of how to speed this up (theoretically) we will need to measure to determine if it's actually faster:

In the current implementation we call `get_secret_parameters()` which fully traverses the entire JSON schema looking for the `secret` property. Next we call `mask_secret_parameters` which traverses the object, but only for the `secret_parameters` passed in.

To potentially speed this up, we could potentially skip traversing parts of the JSON schema if the input data is missing those properties. 

Example: if we have a schema with a nested dictionary

```
{
    'arg_string': {
        'description': 'Junk',
        'type': 'string',
    }
    'arg_optional_object': {
        'description': 'Mirror',
        'type': 'object',
        'properties': {
            'arg_nested_object': {
                'description': 'Mirror mirror',
                'type': 'object',
                'properties': {
                    'arg_double_nested_secret': {
                        'description': 'Deep, deep down',
                        'type': 'string',
                        'secret': True
                    }
                }
            },
            'arg_nested_secret': {
                'description': 'Deep down',
                'type': 'string',
                'secret': True
            }
        }
    }
}
```

And we pass in data:

```
{'arg_string': 'test'}
```

Then we can skip traversing down into the `arg_optional_object` schema since the input data doesn't have this property, saving us time/cycles.

**Comment by nmaludy** [comment]
Additional thing to look at is the `copy.deepcopy()` when calling this recursively we can end up using a ton of memory. Need to figure out a memory-efficient way of doing `mask_secret_parameters()`

**Comment by Kami** [comment]
As far as deep copy goes - yeah, ideally we would create a deep copy of the original object just once and then modify it in subsequent recursive calls (instead of creating a copy during each recursion which as you have said, is problematic).

And in general, I like to avoid recursion where possible because of issues like that, but here it should be fine, if we only create a deep copy once and manipulate that original object in place in subsequent recursive calls.

**Comment by nmaludy** [comment]
@Kami Think this is ready for a review, all unit tests are passing on my local box and with the fix detailed below they should pass in travis.

FYI i ran into a "random failure" in the `st2client/tests/unit/test_inquiry.py` suite where random UUIds were getting generated and causing the `count()` function to return too many values in the CLI list output.

File: st2client/tests/unit/test_inquiry.py [hunk 1]
```diff
@@ -32,6 +32,11 @@
 def _randomize_inquiry_id(inquiry):
     newinquiry = copy.deepcopy(inquiry)
     newinquiry['id'] = str(uuid.uuid4())
+    # ID can't have '1440' in it, otherwise our `count()` fails
+    # when inspecting the inquiry list output for test:
+    # test_list_inquiries_limit()
+    while '1440' in newinquiry['id']:
+        newinquiry['id'] = str(uuid.uuid4())
```
Comment (Kami) [inline st2client/tests/unit/test_inquiry.py]: Thanks for fixing this intermediate test failure :+1:

File: st2common/st2common/util/secrets.py [hunk 2]
```diff
@@ -27,36 +27,146 @@
...
+    # how we iterate depends on what data type was passed in
+    iterator = None
+    is_dict = isinstance(secret_parameters, dict)
+    is_list = isinstance(secret_parameters, list)
+    if is_dict:
+        iterator = six.iteritems(secret_parameters)
+    elif is_list:
+        iterator = enumerate(secret_parameters)
+    else:
+        return MASKED_ATTRIBUTE_VALUE
```
Comment (Kami) [inline st2common/st2common/util/secrets.py]: That's fine or we could also throw because an unsupported type was passed it :)

File: st2common/st2common/util/secrets.py [hunk 3]
```diff
@@ -27,36 +27,146 @@
...
+
+    # only create a deep copy of parameters on the first call
+    # all other recursive calls pass back referneces to this result object
+    # so we can reuse it, saving memory and CPU cycles
+    if not result:
```
Comment (Kami) [inline st2common/st2common/util/secrets.py]: Maybe it's safer to do a more explicit check `if result is None`? Is there a possible scenario where empty dict / array could be returned during recursion?

File: st2common/tests/unit/test_util_secrets.py [hunk 4]
```diff
@@ -0,0 +1,671 @@
...
+                },
+                {
+                    'arg_nested_secret': MASKED_ATTRIBUTE_VALUE
+                },
+                {
+                    'arg_nested_secret': MASKED_ATTRIBUTE_VALUE
+                }
+            ]
+        }
+        self.assertEqual(expected, result)
```
Comment (Kami) [inline st2common/tests/unit/test_util_secrets.py]: Nice work on the extensive tests :+1:

**Comment by Kami** [comment]
Can you please add a corresponding changelog entry (and please try to be explicit and describe scenario where this could occur)?

Besides that, LGTM :+1:

**Comment by Kami** [comment]
The fix for PY3 test failure is in my PR - #4134.

I hope to finish and merge it today.

**Comment by nmaludy** [comment]
OK, i've updated the changelog and converted that one line to `if result is None`. All checks have passed, we should be good to go.

File: CHANGELOG.rst [hunk 5]
```diff
@@ -42,6 +42,18 @@ Changed
 Fixed
 ~~~~~
 
+* Fixed a bug where secrets in pack configs weren't being masked. Recently we
```
Comment (Kami) [inline CHANGELOG.rst]: Minor thing - we use 100 characters wide lines. Not a big deal and this can be changed in a subsequent PR.
